.github/workflows/test.yml

```diff
@@ -0,0 +1,325 @@
+# This is our main "CI tests" workflow. It runs everything that should run on
+# both PRs and merged commits, and for the latter reports failures to slack.
+name: OSS CI
+
+on:
+  push:
+    branches:
+      - "main"
+      - "release-branch/*"
+  pull_request:
+    branches:
+      - "*"
+
+concurrency:
+  # For PRs, later CI runs preempt previous ones. e.g. a force push on a PR
+  # cancels running CI jobs and starts all new ones.
+  #
+  # For non-PR pushes, concurrency.group needs to be unique for every distinct
+  # CI run we want to have happen. Use run_id, which in practice means all
+  # non-PR CI runs will be allowed to run without preempting each other.
+  group: ${{ github.workflow }}-$${{ github.pull_request.number || github.run_id }}
+  cancel-in-progress: true
+
+jobs:
+  test:
+    strategy:
+      fail-fast: false # don't abort the entire matrix if one element fails
+      matrix:
+        include:
+          - goarch: amd64
+          - goarch: amd64
+            race: true
+          - goarch: "386" # thanks yaml
+    runs-on: ubuntu-latest
+    steps:
+    - name: checkout
+      uses: actions/checkout@v3
+    - name: build all
+      run: ./tool/go build ./...
+      env:
+        GOARCH: ${{ matrix.goarch }}
+    - name: build variant CLIs
+      run: |
+        ./build_dist.sh --extra-small ./cmd/tailscaled
+        ./build_dist.sh --box ./cmd/tailscaled
+        ./build_dist.sh --extra-small --box ./cmd/tailscaled
+        rm -f tailscaled
+      env:
+        GOARCH: ${{ matrix.goarch }}
+    - name: get qemu # for tstest/archtest
+      if: matrix.goarch == 'amd64' && !matrix.race
+      run: |
+        sudo apt-get -y update
+        sudo apt-get -y install qemu-user
```

Comment:
I think this is the one which used to be more complicated, pulling some upstream QEMU for Ubuntu 20.04 and using the qemu package for 22.04.

I guess all of GitHub's runners are now 22.04.

Reply:
Yeah, the previous many-files workflow ran on a mix of old ubuntu by explicit version number, newer ubuntu by explicit version number, and ubuntu-latest meaning "whatever github thinks is newest".

For this refactor I switched everything to ubuntu-latest, because if we have any load-bearing dependency on the exact version of ubuntu, it's most likely what you said: we need a recent-enough version of something.